packages/@aws-cdk/aws-cloudwatch/lib/anomaly-detection-alarm.ts

```diff
@@ -0,0 +1,309 @@
+ *
+ * @resource AWS::CloudWatch::Alarm
+ */
+export class AnomalyDetectionAlarm extends AlarmBase {
+
+  /**
+   * ARN of this alarm
+   *
+   * @attribute
+   */
+  public readonly alarmArn: string;
+
+  /**
+   * Name of this alarm.
+   *
+   * @attribute
+   */
+  public readonly alarmName: string;
+
+  /**
+   * The metric object this alarm was based on
+   */
+  public readonly metric: IMetric;
+
+  constructor(scope: Construct, id: string, props: AnomalyDetectionAlarmProps) {
+    super(scope, id, {
+      physicalName: props.alarmName,
+    });
+
+    const comparisonOperator = props.comparisonOperator || AnomalyDetectionComparisonOperator.LESS_THAN_LOWER_OR_GREATER_THAN_UPPER_THRESHOLD;
+
+    const alarm = new CfnAlarm(this, 'Resource', {
+      // Meta
+      alarmDescription: props.alarmDescription,
+      alarmName: this.physicalName,
+
+      // Evaluation
+      comparisonOperator,
+      datapointsToAlarm: props.datapointsToAlarm,
+      evaluationPeriods: props.evaluationPeriods,
+      treatMissingData: props.treatMissingData,
+
+      // Actions
+      actionsEnabled: props.actionsEnabled,
+      alarmActions: Lazy.list({ produce: () => this.alarmActionArns }),
+      insufficientDataActions: Lazy.list({ produce: (() => this.insufficientDataActionArns) }),
+      okActions: Lazy.list({ produce: () => this.okActionArns }),
+
+      // Metric
+      ...this.renderMetric(props.metric, props.threshold),
+    });
+
+    this.alarmArn = this.getResourceArnAttribute(alarm.attrArn, {
+      service: 'cloudwatch',
+      resource: 'alarm',
+      resourceName: this.physicalName,
+      arnFormat: ArnFormat.COLON_RESOURCE_NAME,
+    });
+    this.alarmName = this.getResourceNameAttribute(alarm.ref);
+
+    this.metric = props.metric;
+
+    for (const w of this.metric.warnings ?? []) {
+      Annotations.of(this).addWarning(w);
+    }
+  }
+
+  /**
+   * Trigger this action if the alarm fires
+   *
+   * Typically the ARN of an SNS topic or ARN of an AutoScaling policy.
+   */
+  public addAlarmAction(...actions: IAlarmAction[]) {
+    if (this.alarmActionArns === undefined) {
+      this.alarmActionArns = [];
+    }
+
+    this.alarmActionArns.push(...actions.map(a =>
+      this.validateActionArn(a.bind(this, this).alarmActionArn),
+    ));
+  }
+
+  private validateActionArn(actionArn: string): string {
+    const ec2ActionsRegexp: RegExp = /arn:aws:automate:[a-z|\d|-]+:ec2:[a-z]+/;
+    if (ec2ActionsRegexp.test(actionArn)) {
+      // Check per-instance metric
+      const metricConfig = this.metric.toMetricConfig();
+      if (metricConfig.metricStat?.dimensions?.length != 1 || metricConfig.metricStat?.dimensions![0].name != 'InstanceId') {
+        throw new Error(`EC2 alarm actions requires an EC2 Per-Instance Metric. (${JSON.stringify(metricConfig)} does not have an 'InstanceId' dimension)`);
+      }
+    }
+    return actionArn;
+  }
+
+  private renderMetric(metric: IMetric, threshold: number) {
```

Comment:
question: how much of this is copied directly from `Alarm.renderMetric` and how much is original?

Reply:
It originally was copied directly from `Alarm.reenderMetric`.  I then:

1. Added code to assign/generate an id for the anomaly band expression
2. Added code to resolve the target metric's id, generating it earlier than originally if it isn't assigned in the case of a MathExpression
3. Updated the returns of the callbacks to include the `thresholdMetricId` value that was assigned/generated in the output.
4. Updated the returns of the callbacks for `dispatchMetric` to include the extra metric:
   ```
            {
              expression: `ANOMALY_DETECTION_BAND(${metricId}, ${threshold})`,
              id: thresholdMetricId,
              label: 'Expected',
              returnData: true,
            },
   ```